python/descriptor.c

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright (c) 2009-2021, Google LLC
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of Google LLC nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL Google LLC BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "python/descriptor.h"
+
+#include "python/protobuf.h"
+#include "upb/def.h"
+
+// -----------------------------------------------------------------------------
+// DescriptorBase
+// -----------------------------------------------------------------------------
+
+// This representation is used by all concrete descriptors.
+
+typedef struct {
+  PyObject_HEAD
+  PyObject *pool;  // We own a ref.
+  const void *def;  // Type depends on the class. Kept alive by "pool".
+} PyUpb_DescriptorBase;
+
+PyObject *PyUpb_AnyDescriptor_GetPool(PyObject *desc) {
+  PyUpb_DescriptorBase *base = (void *)desc;
+  return base->pool;
+}
+
+static PyObject *PyUpb_DescriptorBase_New(PyTypeObject *subtype, PyObject *args,
+                                          PyObject *kwds) {
+  return PyErr_Format(PyExc_RuntimeError,
+                      "Creating descriptors directly is not allowed.");
+}
+
+static PyObject *PyUpb_DescriptorBase_NewInternal(PyTypeObject *type,
+                                                  const void *def,
+                                                  PyObject *pool) {
+  PyUpb_DescriptorBase *base = (PyUpb_DescriptorBase *)PyUpb_ObjCache_Get(def);
+
+  if (!base) {
+    base = PyObject_New(PyUpb_DescriptorBase, type);
+    base->pool = pool;
+    base->def = def;
+    Py_INCREF(pool);
```

Comment:
you use Py_INCREF here indicating it is not null and Py_CLEAR below.  feels like one of these is incosnstient

Reply:
Done.